src/sdk/pynni/nni/pbt_tuner/pbt_tuner.py

```diff
@@ -0,0 +1,337 @@
+        for i, x_i in enumerate(x):
+            if isinstance(x_i, dict):
+                if NodeType.NAME not in x_i.keys():
+                    raise RuntimeError('\'_name\' key is not found in this nested search space.')
+            y += json2space(x_i, oldy[i] if oldy else None, name + "[%d]" % i)
+    return y
+
+
+def json2parameter(x, is_rand, random_state, oldy=None, Rand=False, name=NodeType.ROOT):
+    """
+    Json to paramaters.
+
+    """
+    if isinstance(x, dict):
+        if NodeType.TYPE in x.keys():
+            _type = x[NodeType.TYPE]
+            _value = x[NodeType.VALUE]
+            name = name + '-' + _type
+            Rand |= is_rand[name]
+            if Rand is True:
+                if _type == 'choice':
+                    _index = random_state.randint(len(_value))
+                    y = {
+                        NodeType.INDEX: _index,
+                        NodeType.VALUE: json2parameter(
+                            x[NodeType.VALUE][_index],
+                            is_rand,
+                            random_state,
+                            None,
+                            Rand,
+                            name=name+"[%d]" % _index
+                        )
+                    }
+                else:
+                    y = getattr(parameter_expressions, _type)(*(_value + [random_state]))
+            else:
+                y = copy.deepcopy(oldy)
+        else:
+            y = dict()
+            for key in x.keys():
+                y[key] = json2parameter(
+                    x[key],
+                    is_rand,
+                    random_state,
+                    oldy[key] if oldy else None,
+                    Rand,
+                    name + "[%s]" % str(key)
+                )
+    elif isinstance(x, list):
+        y = list()
+        for i, x_i in enumerate(x):
+            if isinstance(x_i, dict):
+                if NodeType.NAME not in x_i.keys():
+                    raise RuntimeError('\'_name\' key is not found in this nested search space.')
+            y.append(json2parameter(
+                x_i,
+                is_rand,
+                random_state,
+                oldy[i] if oldy else None,
+                Rand,
+                name + "[%d]" % i
+            ))
+    else:
+        y = copy.deepcopy(x)
+    return y
+
+
+def exploit_and_explore(bot_trial_info, top_trial_info, factors, epoch):
+    """
+    Replace checkpoint of bot_trial with top, and perturb hyperparameters
+
+    """
+    bot_checkpoint_dir = bot_trial_info.checkpoint_dir
+    top_hyper_parameters = top_trial_info.hyper_parameters
+    hyper_parameters = copy.deepcopy(top_hyper_parameters)
+    for key in hyper_parameters.keys():
+        if key=='load_checkpoint_dir':
+            hyper_parameters[key] = hyper_parameters['save_checkpoint_dir']
+        elif key=='save_checkpoint_dir':
+            hyper_parameters[key] = os.path.join(bot_checkpoint_dir, str(epoch))
```

Comment:
`bot_checkpoint_dir` may has included epoch, it would be not correct to append epoch again.